pkg/utils/kubernetes/reduce/filters.go

```diff
@@ -94,24 +94,56 @@ func filterClusterRoles(clusterRoles []rbacv1.ClusterRole) []rbacv1.ClusterRole
 	return append(clusterRoles[:best], clusterRoles[best+1:]...)
 }
 
+// -----------------------------------------------------------------------------
+// Filter functions - Roles
+// -----------------------------------------------------------------------------
+
+// filterRoles filters out the Role to be kept and returns
+// all the Roles to be deleted.
+// The filtered-out Role is decided as follows:
+//  1. creationTimestamp (newer is better, because newer Roles can contain new policy rules)
+func filterRoles(roles []rbacv1.Role) []rbacv1.Role {
+	if len(roles) == 1 {
+		return []rbacv1.Role{}
+	}
+
+	best := 0
+	for i, cr := range roles {
+		if cr.CreationTimestamp.After(roles[best].CreationTimestamp.Time) {
+			best = i
+		}
+	}
+
+	return append(roles[:best], roles[best+1:]...)
```

Comment:
There is a stdlib helper for this
```suggestion
	return slices.Delete(roles, best, best+1)
```
But, still modifying the slice passed as an argument may lead to side effects that manifest in the code when the argument and/or returned slice is reused later and modified. 

So the bullet-proof solution is to use 

```go 
return slices.Concat(roles[:best], roles[best+1:])
```

which will allocate a new slice, but we have a dedicated issue to maybe enhance everywhere
- https://github.com/Kong/gateway-operator/issues/146 

> This comment is rather to acknowledge a fact than a call to action